hir-type/src/lib.rs

```diff
@@ -662,6 +675,46 @@ impl fmt::Display for LiftedFunctionType {
     }
 }
 
+/// The Miden function type as it is defined in the MASM source (tx kernel, stdlib)
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MidenAbiFunctionType {
```

Comment:
I'd _really_ like to avoid adding yet another copy of `FunctionType` in the IR. Especially since in this case, `MidenAbiFunctionType` conveys no new semantics. The more uniform the IR, the better - we should only add new types for things when there is a strong rationale for doing so.

I think this makes it evident to me that we should extend `FunctionType` with the notion of ABI, and remove this and `LiftedFunctionType`. Conveying the ABI is basically the only reason for `LiftedFunctionType` to exist at all, and `MidenAbiFunctionType` makes it clear we need it for more than just a single ABI.

I think we should add a new `Abi` enum to `hir-type`, with the following values:

* `Wasm`, i.e. the type signature of a function expressed in terms of the _Canonical ABI_ of the Wasm Component Model. This indicates that additional lowering/lifting code is required between caller and callee. It also dictates the calling convention for the callee.
* `Canonical`, i.e. the type signature of a function in canonical form. This is the default ABI unless otherwise specified. Canonical in this context means that no special lowering is required between caller and callee - all the compiler needs to deal with are the details of the specific calling convention.
* `Miden`, i.e. the type signature of a procedure in Miden Assembly. Similarly to `Wasm`, this ABI indicates that the additional lifting/lowering code is required between caller and callee, and dictates the calling convention for the callee.

Once defined, add an `abi` field to `FunctionType`, and add the following to the `FunctionType` API:

```rust
pub fn new_with_abi(abi: Abi, ...) -> Self;
pub fn with_abi(mut self, abi: Abi) -> Self;
pub fn abi(&self) -> Abi;
```

Then we can get rid of the duplicate `FunctionType` clones, and use a single type to represent all these various use cases.

Thoughts?

Reply:
I agree. I might've carried away a bit with creating types for every ABI.
Done in https://github.com/0xPolygonMiden/compiler/pull/140/commits/dd340f0408baa1c2b17423e87f455017eb8da5a0 
I made `FunctionType::new` to set `Abi::Canonical` and added `new_wasm` and `new_miden` for ergonomics.

Reply:
No worries at all! I understand the inclination to do so, especially since the exact shape and semantics of these things has been up in the air - easier to work with new types that don't require modifying a bunch of downstream code. I think up until now, it felt like maybe a new type for the Component Model bits made sense to both of us, but this sort of cemented the idea that we were just fundamentally missing something in `FunctionType`. I think that's an ideal way to evolve our APIs anyway 🙂